Add support for guests to see and subscribe web-public streams.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
Commit summary -
- Backend code to allow guests to subscribe themselves to web-public streams.
- Add `web_public_streams_enabled_for_realm` helper in frontend.
- Pass include_web_public as True for guests to do_get_streams.
- Enable `All streams` tab for guests if web-public streams are enabled at server and realm level.
- Show unsubscribed web-public streams in UI if web-public streams are enabled at server and realm level.
- Allow guests to subscribe themselves to web-public streams from UI.
- Show `Subscribe to more streams` link for guests only if web-public streams are enabled at server and realm level.
- Show "+" icon to guests only if web-public streams are enabled at server and realm level.
Fixes: <!-- Issue link, or clear description.--> #14490

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![web-public](https://user-images.githubusercontent.com/35494118/166116519-aa06c289-93a6-4d71-b143-9e72248417ea.gif)


**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

sahil839 committed:
streams: Enable "All streams" tab for guests if web-public streams are enabled.

sahil839 committed:
streams: Show unsubscribed web-public streams to guest in UI.

sahil839 committed:
left_sidebar: Show "Subscribe to more streams" link to guest.

We show "Subscribe to more streams" link to guest users if
web-public streams are enabled in realm.

sahil839 committed:
streams: Allow guest to subscribe themselves to web-public streams.

This commit adds backend code to allow guests to subscribe
themselves to web-public streams. We remove the require_non_guest_user
decorator from the relevant view function and add separate conditions
to check whether guest can do other actions like adding other users
to streams or create streams using that same view function.
In this commit, we keep the error message returned same as before,
though it can be changed but we can do it after further discussion.

sahil839 committed:
streams: Allow guests to subscribe to web-public streams from UI.

This commit changes the frontend code to allow guests to subscribe
themselves to web-public streams from webapp.

sahil839 committed:
settings_data: Add web_public_streams_enabled_for_realm helper.

sahil839 committed:
left_sidebar: Show "+" only if user is allowed to join or create streams.

We show "+" icon in left sidebar always for non-guest users since they
are always allowed to atleast join streams. And for guest users we show
the icon only when web-public streams are enabled irrespective of
whether there are any web-public streams to join or not.

Fixes #14490.

sahil839 commented on zerver/views/streams.py:
```diff
@@ -532,6 +526,9 @@ def add_subscriptions_backend(
     )
     if len(unauthorized_streams) > 0 and authorization_errors_fatal:
+        if user_profile.is_guest:
+            raise JsonableError(_("Not allowed for guest users"))
```
I am not sure what error message would be best here and below after `if len(principals) > 0:`, but kept it same as before for now.

sahil839 commented on static/js/settings_data.ts:
```diff
@@ -112,6 +112,18 @@ export function user_can_change_logo(): boolean {
+    }
+
+    return page_params.realm_enable_spectator_access;
+}
```
Other functions in this file are mainly for checking whether a user has permission to do this task, but this is whether realm is configured as such or not, but could not find any other suitable file for this. Also, we have this function in server also but we do not include its value in `/register` response currently. Not sure whether we want to directly use that but seemed simple enough to implement in the client itself.

sahil839 commented:
One place where I still need to make sure that we allow web-public streams query when they are actually enabled is `do_get_streams`, otherwise I think this PR has covered all the cases that weren't previously.

timabbott commented on zerver/views/streams.py:
```diff
@@ -772,6 +773,8 @@ def get_topics_backend(
     if is_web_public_query:
         realm = get_valid_realm_from_request(request)
+        if not realm.allow_web_public_streams_access():
+            raise MissingAuthenticationError()
```
I think probably the right model is to have this check occur inside `access_web_public_stream`; generally for authorization things, one wants to rely on common code as much as possible.

sahil839 replied:
Updated.

alya commented:
Nice!

Playing around with this, I noticed that we incorrectly pop up the "Once you leave this stream, you will not be able to rejoin." warning when a guest unsubscribes from a web-public stream.

alya commented:
Otherwise things look good!

sahil839 committed:
streams: Don't show warning while unsubscribing to guests always.

We now do not show the modal when guests are unsubscribing, when
stream is web-public and web-public streams are enabled in realm.
In all other cases, we show the modal.

sahil839 committed:
events: Pass include_web_public as True for guests to do_get_streams.

We pass include_as_web_public as True for guests to do_get_streams
since guests can have access to unsubscribed web-public streams
as well. We need not pass it True for non-guest users since in that
case web-public streams are already included with public streams.

sahil839 committed:
streams: Send stream creation event for web-public streams to guests.

We now send stream creation event for web-public streams to guest
users also if web-public streams are enabled in realm, which were
previously sent to non-guest users only.

This prevents "Unknown stream ID" errors in webapp shown on live-update
when subscribing guest user to a web-public stream. Note that we still
need to fix these errors when subscribing guest to other types of
streams.

sahil839 commented:
Updated the PR.
- Added commit to not show the warning when guest unsubscribes from web-public streams and web-public streams are enabled.
- Also added commit to pass `include_web_public` as `True` to `do_get_streams` for guests.
- Added commit to send stream-creation events for web-public streams to guests.

We still throw error to guest when some other user subscribes the guest user to some stream. Working on a fix for this,

alya commented:
I did some manual testing, and it looks good to me!

sahil839 commented:
Rebased the PR.

alya commented:
Looks like this one needs to be rebased again.

sahil839 commented:
Rebased the PR for CI to pass.

sahil839 commented:
Just an overall summary of the state of this PR -
- We allow guest user to access unsubscribed web-public streams in webapp-UI and also they can subscribe themseleves to web-public streams.
- Note that API allows guet users access to existing web-public stream and its topics, etc. even when the `enable_spectator_access` setting or server-level setting is disabled. We decided to keep it as it is to avoid complexity. Though, we do not allow users to access them or subscribe to them from webapp UI.
- Also, need to add the code for live-update the webapp UI on updating `enable_spectator_access` setting.

timabbott commented:
We've sorta lost track of this, and it's not a near-term priority, but did want to note that https://github.com/zulip/zulip/pull/26899#discussion_r1486974344 adds a new place we'll need to tweak to complete this.

sahil839 commented:
I will have a look at this if and when I get to this after completing the higher priority projects.

shuklamaneesh23 commented:
@sahil839 , Would you be happy reviewing the PR I build over your PR for this issue.
Thanks!

sahil839 commented:
This is not a priority for 10.0, so will review it once I get done with all the high priority things. Can you just mention me on your PR once, so I can add that to my todo list. It is always good to link the PR while asking for review.

shuklamaneesh23 commented:
@sahil839, I haven’t filed a PR for this issue yet. I just wanted to ask for your permission to work on it by using your PR as a reference and completing the work you started here.

sahil839 commented:
I think you can currently work on other higher priority issues then for now. We are not sure how this would pan out after we introduce the setting controlling access to channels.

timabbott commented:
I think we'll be able to replace this with #33320, but we may find we want to grab some of the test cases.
